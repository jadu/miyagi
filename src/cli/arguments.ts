import Arguments from '../interfaces/Arguments';

/**
 * Convert process arguments to an argument object.
 * Expects arguments to be passed as flags e.g. "--foo bar"
 * @param argv
 */
export function getArguments (
    argv: string[]
): Arguments {
    const args: string[] = argv.splice(2);
    const flagTest = new RegExp(/^--/);
    let activeFlag: string;

    // generate arguments object from flags
    return args.reduce((argumentObj: Arguments, arg: string) => {
        // test for flags
        if (flagTest.exec(arg) !== null) {
            // strip "--"
            const argName = arg.substring(2);
            // set active flag
            activeFlag = argName;
            // create empty argument value string
            argumentObj[argName] = '';
        } else if (activeFlag !== undefined) {
            // concat argument values to the active flag, seperate values with a space
            argumentObj[activeFlag] += (argumentObj[activeFlag].length ? ' ' : '') + arg;
        }

        return argumentObj;
    }, {});
}

export function validateArguments (
    required: string,
    args: Arguments
): Arguments {
    // Create array of argument keys
    const argsObj: string[] = Object.keys(args);
    // Search required arguments
    const errors = required.split(' ').filter((req: string) => argsObj.indexOf(req) === -1);
    // Throw error detailing each error
    if (errors.length) {
        throw new Error(errors.map((e: string) => `Required argument "--${e}" is missing`).join('\n'));
    }

    return args;
}
