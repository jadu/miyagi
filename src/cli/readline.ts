import { ReadLine } from 'readline';

/**
 * Wrap readline.question in a promise
 * @param rl
 * @param question
 */
export function question (
    rl: ReadLine,
    question: string
) {
    return new Promise((resolve) => {
        rl.question(question, resolve);
    });
}
