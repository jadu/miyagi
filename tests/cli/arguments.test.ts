import { getArguments, validateArguments } from '../../src/cli/arguments';

describe('getArguments', () => {
    test('should create an arguments object from cli flags', () => {
        const args: string[] = ['/usr/bin/node', 'foo.js', '--foo', 'bar', '--test', 'multiple', 'args'];

        expect(getArguments(args)).toEqual({
            foo: 'bar',
            test: 'multiple args'
        });
    });

    test('should ignore arguments that are not defined as flags', () => {
        const args: string[] = ['/usr/bin/node', 'foo.js', 'bar', '--test', 'multiple', 'args'];

        expect(getArguments(args)).toEqual({
            test: 'multiple args'
        });
    });
});

describe('validateArguments', () => {
    test('should throw an error for each argument that is missing', () => {
        const args = {};

        expect(validateArguments.bind(null, 'foo bar', args)).toThrowErrorMatchingSnapshot();
    });

    test('should not throw an error if arguments are preset', () => {
        const args = {
            foo: '',
            bar: ''
        };

        expect(validateArguments('foo bar', args)).toBe(undefined);
    });
});
