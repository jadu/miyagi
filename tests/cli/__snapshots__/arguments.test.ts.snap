// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateArguments should throw an error for each argument that is missing 1`] = `
"Required argument \\"--foo\\" is missing
Required argument \\"--bar\\" is missing"
`;
